### DISCUSSION TITLE
How do you use a constant to size an array event argument

### DISCUSSION THREAD

**Discussion by JackNWhite** [nasa/fprime#1509] [Q&A] [answered]
My F' instance is interfacing with external code which is the authority on the size of a particular array, and therefore also the size of the telemetry packet used to relate it to ground.

I want to do this:
```xml
<event id="9"
       name="MY_ARRAY"
       severity="ACTIVITY_HI"
       format_string="%s">
    <comment>The thing of the jobbler.</comment>
    <args>
        <arg name="my_array" type="I32" size="MY_ALREADY_DEFINED_ARRAY_LENGTH" />
    </args>
</event>
```
... where `MY_ALREADY_DEFINED_ARRAY_LENGTH` is a preprocessor constant.

When generating the code, however, I get this difficulty:
```
ERROR:RELAXNGV:RELAXNG_ERR_INTEREXTRA: Extra element args in interleave
ERROR:RELAXNGV:RELAXNG_ERR_INTEREXTRA: Extra element events in interleave
/path/to/MyComponentAi.xml:1:0:ERROR:RELAXNGV:RELAXNG_ERR_CONTENTVALID: Element component failed to validate content
```

What is the remedy here?

**Comment by timcanham** [comment, marked as answer]
The XML expects an integer for the size in the attribute, so it fails validation. That explains the error. The XML specification doesn't allow for references to external code, so you'll have to manually keep them in sync.

**Reply by JackNWhite** [reply]
Can I make it a constant within F', so I can define my external constant against that?

**Reply by timcanham** [reply]
The old XML doesn't have constants, but the new FPP does. @bocchino: Does an FPP constant end up in a header somewhere that can be used by C++ code?

**Reply by bocchino** [reply]
Yes, each F Prime build module that defines FPP constants should have a generated file `FppConstantsAc.hpp` that you can include. See https://fprime-community.github.io/fpp/fpp-users-guide.html#Analyzing-and-Translating-Models_Generating-C-Plus-Plus_Constant-Definitions.
